Improve "Cannot read .cabal file inside ..." errors

9999years opened the pull request:
This error message was very confusing -- it doesn't tell you what `.cabal` file it's looking for or why it cannot read the `.cabal` file. (Did it fail to parse the `.tar` archive itself? Did parsing the `.cabal` file fail? If so, why?)

This patch improves the error message to include the name of the `.cabal` file being searched for. Additionally, parse errors and warnings are printed, as are format failures in the tarball itself.

I ran into this error while I was writing tests for Cabal and it confused the hell out of me; this is an expanded version of the changes I made to help debug that failure.

Before:

```
Error: [Cabal-7046]
Cannot read .cabal file inside my-lib-1.0.tar.gz
```

After:

```
Error: In <ROOT>/cabal.dist/repo/my-lib-1.0.tar.gz: Errors encountered when parsing cabal file my-lib-1.0/my-lib.cabal:

my-lib-1.0/my-lib.cabal:4:22: error:
unexpected Unknown SPDX license identifier: 'puppy' 

    3 | version:        1.0
    4 | license:        puppy license :)
      |                      ^

my-lib-1.0/my-lib.cabal:5:1: warning:
Unknown field: "puppy"

    4 | license:        puppy license :)
    5 | puppy:          teehee!
      | ^
Error: [Cabal-7046]
Failed to read my-lib-1.0/my-lib.cabal from archive <ROOT>/cabal.dist/repo/my-lib-1.0.tar.gz
```

- [x] Tests have been added (Ask for help if you don’t know how to write them)
  - [ ] Manual QA notes have been added
- [x] A changelog entry has been added in `changelog.d/pr-YOUR_PR_NUMBER`
- [ ] Documentation has been updated
- [x] Haddock comments for new top-level definitions have been added
- [ ] `base` and third-party library imports use qualified imports or explicit import lists
- [x] Commit messages are formatted nicely
- [ ] Post-merge: Backports for older Cabal release branches have been created

geekosaur reviewed (changes requested):
Can we have (a) a PR template and (b) a changelog entry, since this is user visible? Otherwise, nice.

9999years commented on cabal-install/src/Distribution/Client/ProjectConfig.hs:
```diff
@@ -1615,7 +1616,7 @@ readSourcePackageCabalFile verbosity pkgfilename content =
     (warnings, Right pkg) -> do
       unless (null warnings) $
-        info verbosity (formatWarnings warnings)
+        warn verbosity (formatWarnings warnings)
```
Does anyone know why this wasn't `warn` already?

9999years commented on cabal-testsuite/src/Test/Cabal/Prelude.hs:
```diff
@@ -798,17 +803,31 @@ recordMode mode = withReaderT (\env -> env {
     assertFailure $ "unexpected: " ++ needle
   where output = resultOutput result
 
+assertOutputMatches :: MonadIO m => WithCallStack (String -> Result -> m ())
```
Added a couple test helpers.

ulysses4ever reviewed (approved):
Nice!

9999years commented:
OK, added a changelog entry and updated the test suite. I think moving `.cabal` file warnings from `info` to `warn` resulted in a bunch of output being added to the `cli-suite`.

jasagredo commented on cabal-testsuite/src/Test/Cabal/OutputNormalizer.hs:
```diff
@@ -46,6 +46,24 @@ normalizeOutput nenv =
   . resub (posixRegexEscape (normalizerCanonicalTmpDir nenv) ++ sameDir) "<ROOT>/"
+  . (if buildOS == Windows
+        then
+          -- OK. Here's the deal. In `./Prelude.hs`, `withRepoNoUpdate` sets
+          -- `repoUri` to the tmpdir but with backslashes replaced with
+          -- slashes. This is because Windows treats backslashes and forward
+          -- slashes largely the same in paths, and backslashes aren't allowed
+          -- in a URL like `file+noindex://...`.
+          --
+          -- But that breaks the regexes above, which expect the paths to have
+          -- backslashes.
```
The problem mentioned here is in fact true, but that is not the core of the issue. The core of the issue is that `network-uri` cannot parse uris `file+noindex:///C:/...` and this is because file URIs were not defined in the RFC that describes such library ([see](https://www.rfc-editor.org/rfc/rfc3986)).

File uris were described in a later RFC [link](https://www.rfc-editor.org/rfc/rfc8089.html) and in particular Appendix E.2 describes how to parse paths that contain drive letters on Windows.

@hasufell made an excellent job in implementing `file-uri` precisely for this [link](https://hackage.haskell.org/package/file-uri), but we don't use that one. 

We instead use `network-uri` and we use two slashes, so the following:
```
file+noindex://C:/blah
```
results in `C:` being the Auth part of the URI, and `/blah` the actual path. This can be seen in the test pipeline you shared on Matrix, as indeed it mentions paths like `/msys64/tmp/...` which are missing the drive letter.

So in particular we should do the following:
1. When creating a test `local+noindex` repo, use three slashes always. Linux does use three slashes (`"file+noindex://" ++ path` has three slashes), but Windows doesn't.
2. When running `postProcessRepo`, if the scheme is `file+noindex` (a local repo), defer to `file-uri` to resolve it.

There are two other changes needed for the tests to pass:
- an erroneous call to `System.FilePath.Posix.isAbsolute` that should use `System.FilePath.isAbsolute` instead
- When printing the path to the tarFile make sure it uses the right path separator

Then the tests pass without this hacky conversion on the OutputNormalizer.

I am preparing a branch with these.

ulysses4ever replied:
Oh, I remember this issue was litigated to death here: https://github.com/haskell/cabal/pull/9540 URIs is such a cursed topic...

jasagredo reviewed (changes requested):
I am afraid I have to request changes until the comment below is taken care of. I will push a branch that implements the changes I describe, to merge into this one.

jasagredo commented:
I'm sorry it seems like it needs a formatting change. Here is a patch to apply 
[fmt.patch](https://github.com/user-attachments/files/18204190/fmt.patch)

jasagredo reviewed (approved):
Thanks for merging the PR.

jasagredo commented:
Oof we need to add `file-uri` to the bootstrap jobs. That is kind of easy on Linux and a nightmare on Windows, so if someone else could take care of that it would be awesome.

jasagredo commented:
Aha I see those errors and they are because of `file-uri` not supporting uri fragments.
```
ghci> parseFileURI ExtendedPosix "file:/tmp/bar#shared-cache"
Left "endOfInput"
ghci> parseFileURI ExtendedPosix "file:/tmp/bar"
Right (FileURI {fileAuth = Nothing, filePath = "/tmp/bar"})
```
So I guess we will have to split the string? `takeWhile (/= '#')`?

jasagredo commented:
Here is a patch I would push for this change. 
[frag.patch](https://github.com/user-attachments/files/18204453/frag.patch)

9999years commented:
@jasagredo I'm going to undo your changes for now, let this get merged with the hack, and then we can see about making it cleaner in a follow-up PR.

geekosaur commented:
/cc: @9999years

geekosaur commented:
You had the bad luck of being in the queue behind another PR, so it needs to be rebased again.
